### PR TITLE
[Bundle 7/7] validation: Farewell, global Chainstate!

### PR DESCRIPTION
**Pull request by dongcarl**
Based on:  #21767

à la Mr. Sandman
```
Mr. Chainman, bring me a tip (bung, bung, bung, bung)
Make it the most work that I've ever seen (bung, bung, bung, bung)
Rewind old tip till we're at the fork point (bung, bung, bung, bung)
Then tell it that it's time to call Con-nectTip

Chainman, I'm so alone (bung, bung, bung, bung)
No local objects to call my own (bung, bung, bung, bung)
Please make sure I have a ref
Mr. Chainman, bring me a tip!
```

This is the last bundle in the #20158 series. Thanks everyone for their diligent review.
I would like to call attention to https://github.com/bitcoin/bitcoin/issues/21766, where a few leftover improvements were collated.

- Remove globals:
  - `ChainstateManager g_chainman`
  - `CChainState& ChainstateActive()`
  - `CChain& ChainActive()`
- Remove all review-only assertions.

### CODE REVIEW

**Comment by Sjors** [comment]
In the description of a839078:
> 1. It fixes the erroneous use of Ensure*() in rest.cpp, which cause
   crashes in REST contexts.

Can you elaborate on this?

**Comment by dongcarl** [comment]
@Sjors `EnsureAnyChainman` uses the `src/rpc` way of returning errors, by doing:

https://github.com/bitcoin/bitcoin/blob/59869704c0fe0f105d9cbc193e3670917c618b83/src/rpc/blockchain.cpp#L85

However from my understanding, in `src/rest`, the way to return errors is to call `RESTERR`, which will fill the `req` in-out param with the right HTTP error.

In other words, throwing a `JSONRPCError` in the context of `src/rest` is meaningless, and therefore might crash the node much like an `assert` would.

(btw, that commit is part of Bundle 6)

**Comment by Sjors** [comment]
That makes sense, alright, I'll check out episode 6 first.

**Comment by dongcarl** [comment]
Pushed 8135fd659bd1f068aecf4d0ad3f3862127bb0886 -> 509264117c30704c2f048c702aa0ee9ac9657e7a
- Rebased over merge of #21767

File: src/node/interfaces.cpp [hunk 1]
```diff
@@ -217,22 +207,12 @@ class NodeImpl : public Node
...
-        {
-            // TODO: Temporary scope to check correctness of refactored code.
-            // Should be removed manually after merge of
-            // https://github.com/bitcoin/bitcoin/pull/20158
-            LOCK(::cs_main);
-            active_chainstate = &m_context->chainman->ActiveChainstate();
-            assert(std::addressof(::ChainstateActive()) == std::addressof(*active_chainstate));
-        }
-        return active_chainstate->IsInitialBlockDownload();
+        return m_context->chainman->ActiveChainstate().IsInitialBlockDownload();
```
Comment (ariard) [inline src/node/interfaces.cpp:215]: I think you can directly rely on `NodeImpl::chainman()` method here.

**Review by ariard — COMMENTED** [review]
Approach ACK 5092641

**Comment by dongcarl** [comment]
Pushed 509264117c...cc8cc9c4a2
- Addressed https://github.com/bitcoin/bitcoin/pull/21866#discussion_r647844242

File: src/node/interfaces.cpp (hunk 1 above)
Comment (dongcarl) [reply src/node/interfaces.cpp]: Nice catch!

**Comment by jnewbery** [comment]
utACK cc8cc9c4a24bb638baa733c129a87346c66821ba

**Comment by jnewbery** [comment]
This is a trivial rebase past #22084 (consecutive lines touched)

**Comment by jnewbery** [comment]
Strong concept ACK for:

> Over time, we should probably move these mutable global state variables
into ChainstateManager or CChainState so it's easier to reason about
their lifecycles.

I'd be happy to review any of those changes. Please tag me in PRs that implement those.

**Comment by dongcarl** [comment]
Pushed cc8cc9c4a24bb638baa733c129a87346c66821ba...6f994882deafe62e97f0a889d8bdb8c96dcf913d
- Rebased on top of master

-----

> I'd be happy to review any of those changes. Please tag me in PRs that implement those.

@jnewbery will do! 😁 I'm cleaning up my old branch right now which conflicts with Marco's `node/blockstorage` changes...

**Review by jamesob — APPROVED** [review]
ACK cc8cc9c4a24bb638baa733c129a87346c66821ba ([`jamesob/ackr/21866.1.dongcarl.bundle_7_7_validation_f`](https://github.com/jamesob/bitcoin/tree/ackr/21866.1.dongcarl.bundle_7_7_validation_f))

Many scripted-diffs make this an easy, mechanical change. All the ActiveTip() nits can be probably be addressed by doing a single scripted-diff cleanup that seds `'ActiveChain\(\).Tip\(\)'`.

Nice job with all the scripted-diffs; they make review much easier. Man oh man that's a lot of asserts! I like your move of `UnloadBlockIndex()` into the ChainstateManager destructor.

Nice work on this! Who ever thought we'd have chainstate deglobalized?? What a world.


<details><summary>Show signature data</summary>
<p>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK cc8cc9c4a24bb638baa733c129a87346c66821ba ([`jamesob/ackr/21866.1.dongcarl.bundle_7_7_validation_f`](https://github.com/jamesob/bitcoin/tree/ackr/21866.1.dongcarl.bundle_7_7_validation_f))

-----BEGIN PGP SIGNATURE-----

iQIzBAEBCgAdFiEEGNRVI1NPYuZCSIrGepNdrbLETwUFAmDCobMACgkQepNdrbLE
TwX7NhAAgxLYZY+ecZjwI2y5OmqMrOjtmX+SplLm8B+iPpJCfYRkMfqZPG5qaVJQ
7eHCPBP6/V2Rx0CYgMFqX6G+3tJsW6k/ErahtuZXcp8nUEtyaOnt9jbDhdWRqAT9
MS0sEmvw/HNqz3qBpDW0aZs2OO6D01HUi2SkuCbze7zWiZYiAeeInUu1Du7tO0IQ
c3VqzJqz5XRVi64VF0fR5NeDgzgNYMTDWXEnwDK40JWB1gpM9uht5KuRGPLLfPl+
p7gDafWxoAbu9csPmKY9YIgPcQK2yKvQ5DSyWJVm57a0wHxlLFjGMCoprg7kbVn1
jaVyZ4EgghKT/NME6yhAeLBg8HAyaXNYaXISAlVyUM6FbvJ3p3sNpkgCbldz2xTH
nhiibFpwJbQVO/EoWw7Oy8TTe9FJGqlrhpXTR3vdsMy7g6gLwIzRQ71h43UVENs4
x4HnXfMEFPnIvt0W5xsSv6Ti+gWIB1e/nF/EdEwkmXD6s/6llqE7Bs1dGfDqb3jG
S9+2XsPy8KLgn+r3zSH66xKII70NdpUyjvQ9uNki8tVcTa5Jy905RgLxuldT2wrv
CGoazisooMRUfC44+a4Zn/w9IsxPzkUXP6RjmI9HfxF2f2vHWmv1IePwYJs0lD8U
M8x2upnRDRQe5QQn0+geIjsar4uRiEBphx6xBpSkFsWoDMmcP5Q=
=XPto
-----END PGP SIGNATURE-----

```

</p></details>

<details><summary>Show platform data</summary>
<p>

```
Tested on Linux-4.19.0-16-amd64-x86_64-with-glibc2.28

Configured with ./configure LDFLAGS=-L/home/james/src/bitcoin/db4/lib/ CPPFLAGS=-I/home/james/src/bitcoin/db4/include/ CXXFLAGS=-fPIE -pipe -O2 -g -Wthread-safety-analysis -Wall -Werror=sign-compare -Wsign-compare -Werror=thread-safety-analysis --enable-wallet --enable-debug --with-daemon --enable-natpmp-default

Compiled with /usr/bin/ccache /usr/local/bin/clang++ -std=c++17 -mavx -mavx2 -fPIE -pipe -O2 -g -Wthread-safety-analysis -Wall -Werror=sign-compare -Wsign-compare -Werror=thread-safety-analysis -O0 -g3 -ftrapv -fdebug-prefix-map=$(abs_srcdir)=.  -Wstack-protector -fstack-protector-all -fcf-protection=full -fstack-clash-protection -msse4 -msha -msse4.1 -msse4.2  i

Compiler version: Debian clang version 11.1.0-++20210405104510+1fdec59bffc1-1~exp1~20210405085125.161
```

</p></details>

**Comment by jamesob** [comment]
reACK https://github.com/bitcoin/bitcoin/pull/21866/commits/6f994882deafe62e97f0a889d8bdb8c96dcf913d based on the contents of 
`git range-diff master ackr/21866.1.dongcarl.bundle_7_7_validation_f ackr/21866.2.dongcarl.bundle_7_7_validation_f` and local build/test run.

**Comment by ariard** [comment]
Code Review ACK 6f99488.

Diff since last review is using `NodeImpl::chainman` method in `isInitialBlockDownload()` and rebase changes from #22084.

**Comment by jnewbery** [comment]
utACK 6f994882deafe62e97f0a889d8bdb8c96dcf913d

**Comment by achow101** [comment]
Code Review ACK 6f994882deafe62e97f0a889d8bdb8c96dcf913d

**Review by ryanofsky — APPROVED** [review]
Code review ACK 6f994882deafe62e97f0a889d8bdb8c96dcf913d.

Just want to note so it isn't forgotten that 972c5166ee685447a6d4bf5e501b07a0871fba85 commit message suggests some good future followup "Since now ChainstateManager is no longer a global, we can just put this [other variable cleanup] logic in its destructor to make sure that callers are always correct" which would make shutdown more comprehensible, and clean up the Qt tests

**Comment by MarcoFalke** [comment]
Just noting that commit 972c5166ee685447a6d4bf5e501b07a0871fba85 causes a segmentation fault because the commits are in the wrong order:

```
$ valgrind ./src/qt/test/test_bitcoin-qt 
==1844015== Memcheck, a memory error detector
==1844015== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==1844015== Using Valgrind-3.17.0 and LibVEX; rerun with -h for copyright info
==1844015== Command: ./src/qt/test/test_bitcoin-qt
...
********* Finished testing of RPCNestedTests *********
terminate called after throwing an instance of 'std::logic_error'
  what():  should not be overwriting a chainstate
==1844015== 
==1844015== Process terminating with default action of signal 6 (SIGABRT): dumping core
==1844015==    at 0x651A2A2: raise (in /usr/lib64/libc-2.33.so)
==1844015==    by 0x65038A3: abort (in /usr/lib64/libc-2.33.so)
==1844015==    by 0x61FEA45: ??? (in /usr/lib64/libstdc++.so.6.0.29)
==1844015==    by 0x620A29B: ??? (in /usr/lib64/libstdc++.so.6.0.29)
==1844015==    by 0x620A306: std::terminate() (in /usr/lib64/libstdc++.so.6.0.29)
==1844015==    by 0x620A5FC: __cxa_rethrow (in /usr/lib64/libstdc++.so.6.0.29)
==1844015==    by 0x4F392CD: ??? (in /usr/lib64/libQt5Test.so.5.15.2)
==1844015==    by 0x4F43B3F: QTest::qExec(QObject*, int, char**) (in /usr/lib64/libQt5Test.so.5.15.2)
==1844015==    by 0x196413: main (test_main.cpp:91)
==1844015== 
==1844015== Process terminating with default action of signal 11 (SIGSEGV)
==1844015==  General Protection Fault
==1844015==    at 0x4877C82: ??? (in /usr/lib64/libpthread-2.33.so)
==1844015==    by 0x660D03E: ??? (in /usr/lib64/libc-2.33.so)
==1844015==    by 0x664F60C: ??? (in /usr/lib64/libc-2.33.so)
==1844015==    by 0x664F261: __libc_freeres (in /usr/lib64/libc-2.33.so)
==1844015==    by 0x48351E7: _vgnU_freeres (vg_preloaded.c:74)
==1844015== 
==1844015== HEAP SUMMARY:
==1844015==     in use at exit: 45,331,308 bytes in 21,971 blocks
==1844015==   total heap usage: 128,095 allocs, 106,124 frees, 87,543,385 bytes allocated
==1844015== 
==1844015== LEAK SUMMARY:
==1844015==    definitely lost: 0 bytes in 0 blocks
==1844015==    indirectly lost: 0 bytes in 0 blocks
==1844015==      possibly lost: 22,704 bytes in 95 blocks
==1844015==    still reachable: 45,308,604 bytes in 21,876 blocks
==1844015==                       of which reachable via heuristic:
==1844015==                         stdstring          : 63 bytes in 1 blocks
==1844015==                         newarray           : 424 bytes in 11 blocks
==1844015==                         multipleinheritance: 71,160 bytes in 87 blocks
==1844015==         suppressed: 0 bytes in 0 blocks
==1844015== Rerun with --leak-check=full to see details of leaked memory
==1844015== 
==1844015== Use --track-origins=yes to see where uninitialised values come from
==1844015== For lists of detected and suppressed errors, rerun with: -s
==1844015== ERROR SUMMARY: 3 errors from 3 contexts (suppressed: 0 from 0)
Segmentation fault (core dumped)

**Comment by MarcoFalke** [comment]
Otherwise looks good

**Pull request merged**